### ISSUE TITLE
Image not cleared when polygon entity is removed from dataSource

### ISSUE THREAD

**Issue by rahwang** [AnalyticalGraphicsInc/cesium#5379]
Reported in the forum: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/JgzuftedXi0

When more than one polygon with an image material is added with a non-zero height, removing one of the polygons from the entity collection does not clear the image from the screen.

Here's a Sandcastle example:
http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Hello%20World.html&label=Showcases&gist=a0cfb82ab43474f87afdcdaba339e76d

To reproduce the issue:
- Click "Add first entity" 
- Click "Add second entity"
- Click "Remove second entity"

Expected: Removes the second entity along with the image.
Actual: Doesn't clear the image.

**Comment by mramato** [comment]
Note to whomever fixes this (though I'll probably do it when I have more time).  This looks like a race condition involved with `StaticGeometryPerMaterialBatch.update` vs `StaticGeometryPerMaterialBatch.remove` caused by changing an entity property immediately after removing it (but before the remove has been processed by the visualizers).  Because the item has changed, it's still getting put in a batch even though it has also been removed.

**Comment by ggetz** [comment]
Reported on the forum here: https://groups.google.com/forum/#!searchin/cesium-dev/mittner%7Csort:relevance/cesium-dev/PgFyVrT3_jM/aCqAV30rDgAJ

**Comment by rhamie** [comment]
Hi, I recently ran into the same issue as Dave from the above Google Group and as he says, resetting the length variable after the first loop solves the second loop's array out of bounds issue.  I'm new to working with Cesium so I'm not entirely sure how Dave's issue is encapsulated by the race condition outlined here, but it seems that if there's a chance that the array will be modified (the slice in the first loop) the length variable should be reset regardless.

```
StaticGeometryPerMaterialBatch.prototype.update = function(time) {
    var i;
    var items = this._items;
    var length = items.length;                           <= #1 initial length recorded
    for (i = length - 1; i >= 0; i--) {
        var item = items[i];
        if (item.invalidated) {
            items.splice(i, 1);                          <= #2 actual length may changes
            var updaters = item.updaters.values;
            var updatersLength = updaters.length;                   
            for (var h = 0; h < updatersLength; h++) {
                this.add(time, updaters[h]);
            }
            item.destroy();
        }
    }
    length = items.length;                               <=#5 potential solution, works for my case
    var isUpdated = true;
    for (i = 0; i < length; i++) {                       <= #3 initial length still assumed
        isUpdated = items[i].update(time) && isUpdated;  <= #4 potential error
    }                
    return isUpdated;
 }; 
```

**Comment by hpinkos** [comment]
Fixed this somewhere along the way when I refactored geometry entities.  This will be included in the 1.42 release on March 1st

**Comment by cesium-concierge** [comment]
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/forum/?hl&#x3D;en#!topic/cesium-dev/JgzuftedXi0
https://groups.google.com/forum/#!searchin/cesium-dev/mittner%7Csort

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/issues/5379 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.
---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:
